### PR TITLE
Add ITILFollowup case in Search::addDefaultWhere()

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Following #6116
Internal ref : 16937

Add ITILFollowup case in Search::addDefaultWhere().
This change is made to filter correctly the followups in API::getItems()
Currently only check public/private visiblity, need to also check parent item visibility...

Bonus fix : 
- Fix the canViewItem of ITILFollowup which had an unused condition (comparing an object to the string "Ticket").

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | pending

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
This is ready for review, can you remove the WIP label @cedric-anne  ?

File: inc/search.class.php [hunk 1]
```diff
@@ -3892,6 +3943,44 @@ static function addDefaultWhere($itemtype) {
...
+            // Case 1: can see only owned items
+            $user   = Session::getLoginUserID();
+            $groups = "'" . implode("','", $_SESSION['glpigroups']) . "'";
+
+            // Avoid empty IN ()
+            if ($groups == "''") {
+               $groups = '-1';
+            }
+
+            $subQueryUser = "SELECT `$target` from `$userTable` WHERE `users_id` = '$user'";
```
Comment (cedric-anne) [inline inc/search.class.php]: Conditions are more complex on tickets (see https://github.com/glpi-project/glpi/pull/6150/files#diff-a801c52129f44d50ab6d1db9f049c4b7R3733 )

File: inc/search.class.php [hunk 2]
```diff
@@ -3877,6 +3877,57 @@ static function addDefaultWhere($itemtype) {
             $condition = SavedSearch::addVisibilityRestrict();
             break;
 
+         case 'ITILFollowup':
+            // Filer on is_private
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/search.class.php]: Done.

File: inc/search.class.php [hunk 3]
```diff
@@ -3892,6 +3943,44 @@ static function addDefaultWhere($itemtype) {
...
    }
 
+   /**
+    * Build parent condition for ITILFollowup, used in addDefaultWhere
+    *
+    * @param array $params keys: itemType, target, userTable, groupTable
+    *
+    * @return string
+    */
+   public static function buildITILFollowupParentCondition($params) {
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/search.class.php]: Done.

File: inc/search.class.php [hunk 4]
```diff
@@ -3892,6 +3943,44 @@ static function addDefaultWhere($itemtype) {
...
+      $itemType   = $params['itemType'];
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply inc/search.class.php]: Done.

File: inc/search.class.php (hunk 1 above)
Comment (AdrienClairembault) [reply inc/search.class.php]: Need tickets rights check are live, is this OK for you ?

**Comment by AdrienClairembault** [comment]
@cedric-anne @orthagh Is this ok for you ?

File: inc/itilfollowup.class.php [hunk 5]
```diff
@@ -1073,4 +1073,151 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
+               WHERE `users_id_recipient` = '$user'";
+            $condition .= "OR `items_id` IN ($recipient_query) ";
+         }
+
+         if (Session::haveRight($itemtype, $itemtype::READGROUP)) {
+            // Add tickets where the users is in a requester or observer group
+            // Subquery for requester/observer group
+            $group_query = "SELECT `$target`
+               FROM `$group_table`
+               WHERE `users_id` = '$user' AND type IN ($requester, $obs)";
```
Comment (cedric-anne) [inline inc/itilfollowup.class.php]:
```suggestion
               WHERE `groups_id` IN ($groups) AND type IN ($requester, $obs)";
```

File: inc/itilfollowup.class.php [hunk 6]
```diff
@@ -1073,4 +1073,151 @@ static function processMassiveActionsForOneItemtype(MassiveAction $ma, CommonDBT
...
+               WHERE `users_id` = '$user' AND type = $assign";
+            $condition .= "OR `items_id` IN ($user_query) ";
+         }
+
+         if (Session::haveRight($itemtype, $itemtype::READASSIGN)) {
+            // Add tickets where the users is part of an assigned group
+            // Subquery for assigned group
+            $group_query = "SELECT `$target`
+               FROM `$group_table`
+               WHERE `users_id` = '$user' AND type = $assign";
```
Comment (cedric-anne) [inline inc/itilfollowup.class.php]:
```suggestion
               WHERE `groups_id` IN ($groups) AND type = $assign";
```

File: inc/itilfollowup.class.php (hunk 5 above)
Comment (AdrienClairembault) [reply inc/itilfollowup.class.php]: Good catch, should be fixed now :)

**Pull request merged**